src/System.Text.Json/tests/Serialization/ExtensionDataTests.cs

```diff
@@ -214,6 +260,26 @@ public static void ObjectTree()
             Assert.Equal(3, child.MyOverflow["MyIntMissingChild"].GetInt32());
         }
 
+        [Fact]
+        public static void ExtensionProperty_InvalidDictionary()
+        {
+            ClassWithInvalidExtensionPropertyStringString obj1 = new ClassWithInvalidExtensionPropertyStringString();
+            Assert.Throws<InvalidOperationException>(() => JsonSerializer.ToString(obj1));
+
+            ClassWithInvalidExtensionPropertyObjectString obj2 = new ClassWithInvalidExtensionPropertyObjectString();
+            Assert.Throws<NotSupportedException>(() => JsonSerializer.ToString(obj2));
```

Comment:
We've different exception because we don't support serialization of dictionary with key different than `string` 
https://github.com/dotnet/corefx/blob/a78bd7308555926f88aad3d0ecaaebae0f9204cc/src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs#L400
So exception is raise before `DetermineExtensionDataProperty`

https://github.com/dotnet/corefx/blob/94c5f1384c9b6477d164050de964cf6dcf99b2a9/src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs#L166

Reply:
That's a good point, @MarcoRossignoli. I think there is value in us being consistent here. If we are not supporting a type of collection because it doesn't make sense to, `InvalidOperationException` is fine. If, however, we may support a type in the future, we should throw `UnsupportedExeception`.

**So, here are the cases:**
`Dictionary<string, any TValue>` property - should work (only for serialize)
`Dictionary<string, object>` property - should work
`Dictionary<string, JsonElement>` property - should work
`Dictionary<string, user-defined POCO>` property - should work
`Dictionary<non-string TKey, any TValue>`  - not supported, for now (neither serialize/deserialize)

`Dictionary<string, any TValue>` property with JsonExtensionData - should work (only for serialize)
`Dictionary<string, object>` property with JsonExtensionData - should work
`Dictionary<string, JsonElement>` property with JsonExtensionData - should work
`Dictionary<string, user-defined POCO>` property - should work - should work (only for serialize)
`Dictionary<non-string TKey, any TValue>`  property with JsonExtensionData - not supported, for now (neither serialize/deserialize)

@steveharter, what do you think? In what case would it make sense to throw `InvalidOperationException` instead?

Reply:
That said, this can be addressed separately, especially since we want to get the blocking-partner issue resolved.

Reply:
ok so I won't touch code for now on this.